Tuple conversions mop up.

VSadov opened the pull request:
- Explicit tuple literal conversions are now implemented.
- Conversion objects are stored on BoundConversion nodes and used in lowering, recursively.
- Nullable lifted conversions, when lowered, correctly remember whether there was a cast in the code or not
- Reduced the size of most common conversions. We store Conversions in many bound nodes and now in BoundConversions too.

See fixed bugs for more info -

Fixes:https://github.com/dotnet/roslyn/issues/11804
Fixes:https://github.com/dotnet/roslyn/issues/11934
Fixes:https://github.com/dotnet/roslyn/issues/12064
Fixes:https://github.com/dotnet/roslyn/issues/12067

VSadov commented:
@dotnet/roslyn-compiler - please review

VSadov commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -12852,10 +12852,9 @@ static void Main()
-  .maxstack  3
-  .locals init (System.ValueTuple<int, int> V_0,
-                System.ValueTuple<long, object> V_1)
+  // Code size       65 (0x41)
```
Notice the IL change. This is now classified an explicit tuple literal conversion, and 
since it is a literal conversion (not a conversion from a tuple type) we do not materialize an intermediate tuple and IL is more compact.

There is a subtle difference in the order of sideeffects between evaluation of the tuple arguments and element-wise conversions. We are now consistent with implicit tuple literal conversions here and element conversions are interlaced with argument evaluations. There is no intermediate tuple materialization between argument eval and element conversions. 

We did discuss this at LDM, if you wonder... it is a subtle change though.

VSadov commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:
```diff
@@ -674,20 +634,7 @@ private BoundExpression MakeImplicitConversion(BoundExpression rewrittenOperand,
-                else
-                {
-                    conversion = this._compilation.Conversions.ClassifyConversion(fieldAccess.Type, destElementTypes[i], ref useSiteDiagnostics);
-                }
```
No longer need to classify conversions again at lowering (and risk to do it under different context) since we have conversions stored on the bound node.
Nullable lowering has similar changes and that is more important since in some cases we did not have all the context to redo classifications at lowering correctly.

VSadov commented:
Tag @dotnet/roslyn-compiler

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs:
```diff
@@ -101,33 +279,47 @@ internal bool IsExtensionMethod
+            get
             {
-                _flags = IsExtensionMethodMask;
+                var uncommonData = _uncommonData;
```
What's the purpose of this local?

VSadov replied:
to not fetch it multiple times.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -308,18 +309,19 @@ private static ConversionKind ClassifyImplicitConstantExpressionConversion(Bound
         }
 
-        private ConversionKind ClassifyImplicitTupleLiteralConversion(BoundExpression source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        private Conversion ClassifyImplicitTupleLiteralConversion(BoundTupleLiteral source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
```
Consider renaming "ClassifyImplicitTupleLiteralConversion" to "GetImplicitTupleLiteralConversion", for consistency with "GetImplicitTupleLiteralConversion".

VSadov replied:
If I rename it, it would clash with "Get". 
Get just gets a tuple conversion if exists. Classify may also return a nullable conversion whose underlying conversion is a tuple conversion.
"Get" is also consistent with other abstract methods in ConversionBase  - like GetInterpolatedStringConversion who require an access to a binder.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -329,14 +331,47 @@ private ConversionKind ClassifyImplicitTupleLiteralConversion(BoundExpression so
+            return Conversion.NoConversion;
+        }
+
+        private Conversion ClassifyExplicitTupleLiteralConversion(BoundTupleLiteral source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics, bool forCast)
```
Mostly the same as `ClassifyImplicitTupleLiteralConversion`. Could this be factored?

VSadov replied:
the diff size would seem to be comparable with common parts because of "forCast" in the explicit case. I will keep the methods separate

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -398,6 +433,18 @@ internal static bool HasImplicitConstantExpressionConversion(BoundExpression sou
             Debug.Assert((object)destination != null);
 
+            // NB: need to check for explicit tuple literal conversion before checking for explicit conversion from type
+            //     The same literal may have both explicit typle conversion and explicit tuple literal conversion to the target type.
```
Typo: s/typle/tuple/

VSadov replied:
will need to fix

jcouv commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversions.cs:
```diff
@@ -713,38 +760,69 @@ public override Conversion GetMethodGroupConversion(BoundMethodGroup source, Typ
+        protected override Conversion GetImplicitTupleLiteralConversion(BoundTupleLiteral source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
-            if (source.Kind != BoundKind.TupleLiteral)
+            var arguments = source.Arguments;
```
Consider creating a local for `arguments.Length` as well.

VSadov commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs:
```diff
@@ -14,79 +14,257 @@ namespace Microsoft.CodeAnalysis.CSharp
+        // in uncommon cases an instance of this class is attached to the conversion.
+        private class UncommonData
+        {
+            public UncommonData(
```
This API exists for performance reasons and is a bit unsafe. In many cases you would need to pass some additional data besides kind and accidental omitting some data could lead to subtly incorrect behavior.

I was concerned that with default values it could be too easy to pass incomplete set of data, so I want to force the caller to pass default values explicitly (preferable using named parameters) and think/comment why defaults are acceptable.

It should not be very common to call this directly, so I think it is not a big burden.

VSadov commented on src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/Conversion.cs:
```diff
@@ -101,33 +279,47 @@ internal bool IsExtensionMethod
+                        return uncommonData._conversionMethod;
+                    }
+
+                    var conversionResult = _uncommonData._conversionResult;
```
forgot to use uncommonData here.

jcouv commented:
There are comments that need updating (such as Binder_Expressions.cs line 1835), which mention the issue 11804 and says "currently not supported".

jcouv commented:
Would it make sense to have tests for explicit literal conversions on long or nested tuples too?

Otherwise, LGTM

VSadov commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -1829,16 +1829,6 @@ private BoundExpression BindCastCore(ExpressionSyntax node, BoundExpression oper
                 var tuple = (BoundTupleLiteral)operand;
                 var targetElementTypes = default(ImmutableArray<TypeSymbol>);
 
-                // source does not have a type.
```
removed this and added tests.

VSadov commented:
Ok, I think I have addressed all the concerns and PR is a week old. The change is large, but mostly because of mechanical changes - renames and such. 
I will merge the PR before it runs into merge conflicts. If there are more comments, I will address them in separate changes.
